torch/nested/_internal/ops.py

```diff
@@ -255,19 +267,461 @@ def jagged_binary_pointwise(func, *args, **kwargs):
+        )
+    if attn_mask is not None:
+        # TODO: Figure out whether masks are actually supported for this layout or not
+        if attn_mask.dtype != torch.bool and attn_mask.dtype != query.dtype:
+            raise ValueError(
+                f"Expected attn_mask dtype to be bool or to match query dtype, but got attn_mask.dtype: "
+                f"{attn_mask.dtype}, and query.dtype: {query.dtype} instead."
+            )
+
+
+def _can_use_math_sdpa_jagged(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    attn_mask: Optional[torch.Tensor] = None,
+    dropout_p=0.0,
+    is_causal=False,
+    scale=None,
+):
+    if (
+        not query.is_contiguous()
+        or not key.is_contiguous()
+        or not value.is_contiguous()
+    ):
+        raise ValueError("If inputs are nested tensors they must be contiguous.")
+    if is_causal:
+        raise ValueError(
+            "Nested tensors for query / key are not supported when is_causal=True."
+        )
+
+
+def _select_sdp_backend(query, key, value, attn_mask, dropout, is_causal):
+    if (
+        not flash_sdp_enabled()
+        and not mem_efficient_sdp_enabled()
+        and not math_sdp_enabled()
+    ):
+        return SDPBackend.ERROR
+
+    ordering = (
+        SDPBackend.FLASH_ATTENTION,
+        SDPBackend.EFFICIENT_ATTENTION,
+        SDPBackend.MATH,
+    )
+
+    params = SDPAParams(query, key, value, attn_mask, dropout, is_causal)
+
+    for backend in ordering:
+        if backend == SDPBackend.FLASH_ATTENTION:
+            if can_use_flash_attention(params):
+                return SDPBackend.FLASH_ATTENTION
+        if backend == SDPBackend.EFFICIENT_ATTENTION:
+            if can_use_efficient_attention(params):
+                return SDPBackend.EFFICIENT_ATTENTION
+        if backend == SDPBackend.EFFICIENT_ATTENTION:
+            if math_sdp_enabled():
+                return SDPBackend.MATH
+
+    log.warning("Memory efficient kernel not used because:")
+    can_use_efficient_attention(params, True)
+    log.warning("Flash attention kernel not used because:")
+    can_use_flash_attention(params, True)
+    raise ValueError("No available kernel. Aborting execution.")
+
+
+def _cumulative_and_max_seq_len_nnz(qkv):
+    """
+    This function is used to calculate two pieces of metadata that are needed
+    for use with flash-attention and efficient_attention kernels. They are the
+    cumulative sequence_length over a batch of sequences and the maximum
+    sequence length.
+
+    @return A tuple of cumulative sequence lengths and the maximum sequence
+    length, and the last element in the cumulative_sequence_lengths
+    """
+    if not qkv.is_nested:
+        raise ValueError("QKV must be nested for flash cumulative_seq_len calculation.")
+
+    if qkv.lengths() is None:
+        cumulative_seqlen = qkv.offsets().to(dtype=torch.int32, device=qkv.device)
+        max_seqlen = torch.max(qkv.offsets().diff()).item()
```

Comment:
yeah, I am worried about this too, but given offsets is a tensor, I saw no other way to get this metadata

Reply:
apparently it's used to choose the block size for launching the CUDA kernels and has nothing to do with holes in the tensor (form my conversation with @drisspg yesterday)

Reply:
yeah its not explicitly for choosing the block size but it is metadata that is needed on the host for the initial invocation of the kernel

Reply:
yes but what's it used for in the kernel :)

Reply:
Well its used a few places: 1.) Needed to create the size for the logsumexp tensor. 2.) It defines the upper extent of the seqlens which is in fact used for figure out far we need to iterate into each qkv tensor during the execution of iterative softmax and this is added via launch param. The second could likely be read from an on device ptr but I don't think 1 could be

Reply:
I think caching and passing through each jagged tensor makes the most since, since this value + offsets will be the same for each layer in a transformer per iteration

Reply:
I've added two more cached values. max_seqlen and min_seqlen, which are both needed in the host for checks and kernels.